### ISSUE TITLE
Empty/IsEmpty properties

### ISSUE THREAD

**Issue by nathanwinder** [AArnott/ImmutableObjectGraph#109]
Add support for a static `Empty` property and instance `IsEmpty` property.

```csharp
Person.Empty.IsEmpty == true;
Person.Create().IsEmpty == true;
ReferenceEquals(Person.Create(), Person.Empty) == true;
ReferenceEquals(Person.CreateBuilder().ToImmutable(), Person.Empty) == true;
// This could be argued in either direction, what are your thoughts?
ReferenceEquals(Person.Create(firstName: "Nathan").WithFirstName(null), Person.Empty) == false;
Person.Create(firstName: "Nathan").WithFirstName(null).Equals(Person.Empty);
Person.Create(firstName: "Nathan").WithFirstName(null).IsEmpty == true;
```

**Comment by AArnott** [comment]
Before knowing whether an instance is equal to an empty instance, we'd need #107 completed. 
We also need to know what an Empty instance means.

**Comment by nathanwinder** [comment]
I think "Default" implies an object in a valid state while "Empty" is an incomplete or ambiguous state, its purpose being an alternative to `null`. I assume that "required properties" are required via the `Validate` method which can be skipped in the case of creating the Empty instance because empty is not meant to be a "valid" state. All properties would be initialized to their default values or `Empty` in the case of other Immutable types or `String`. Upon mutation from the empty state validation would be applied.

The presence of validation makes the concept of empty concrete because the invalid state of empty is now something special. I see that the meaning of empty is ambiguous when dealing with an object that does not have required properties, as in the case of `Person.Create()`. It does not have values set but it is in a valid state. Perhaps the solution is:

Person.Empty.IsEmpty == true;
Person.Create().IsEmpty == false;
ReferenceEquals(Person.Create(), Person.Empty) == false;

This would make the "invalid" empty state unambiguous as there is only a single instance, it is only accessible through the static property, and it is only equatable by reference.

Empty should be an opt-in feature because it requires the elimination of nulls. For example, none of the properties should accept null values and any unspecified property values should be set to Empty. Otherwise the behavior of the type becomes inconsistent as demonstrated below:

Person.Create().FirstName == null
Person.Empty.FirstName == String.Empty
Person.Empty.WithLastName("Winder").FirstName == String.Empty

C# 8 should help with this and analyzers "could" be written to enforce these rules when applying this feature.

**Comment by nathanwinder** [comment]
Serialization is another consideration. Empty values could be serialized as `null` or an object graph with IsEmpty set to true. I think that should be left to the serializer to decide. It appears that the intent is for deserialization to be done into a builder tree, this works well because it becomes the responsibility of the builder to then transform any null or empty object graphs into proper Empty reference. That leaves an issue with builder state. If a builder IsEmpty and a property is modified should IsEmpty be set to false? And if IsEmpty is set to true should the builder "clear" its properties? Also, if the deserialized data is invalid having values set and the IsEmpty flag set is it empty or does it throw? What are you thoughts on these issues?

**Comment by nathanwinder** [comment]
I think the precident set for the “IsDefault/Default” by the Immutable Collections library is reason enough to use that naming over Empty/IsEmpty. IsDefault indicates an uninitialized/invalid array which matches the intent here.

**Comment by AArnott** [comment]
`IsDefault` only exists on `ImmutableArray<T>` because it is a struct that *must* be initialized or it will throw NRE. That design was hard to swallow (I was part of it), but we did it because it allowed the .NET JIT to generate much more efficient code in tight loops where the struct was used.

I don't see the point of having a non-null reference type that follows that pattern, and is serialized as null.

**Comment by nathanwinder** [comment]
Yes, the `ImmutableArray<T>` API is a bit awkward and I'm sure you guys languished over that trade-off. Perhaps I'm not thinking about this correctly. But the point of supporting `Default` on a reference type is to allow the developer to specify that an object is in an uninitialized/default/empty state without the risk of NRE. Although the more I think about it maybe it is best to leave this to the new C# 8 Non-Nullable Reference types feature. That would achieve this goal without any additional implementation work for this project.
